### PR TITLE
initial implementation of color as state

### PR DESCRIPTION
**Pull request by wllfaria**
this is my initial attempt to implement colors as state, this works. but I'm totally not sure if everything I did is how you envisioned this so I'm letting this here so you can code review and suggest things. But here is how it looks:

image

a decision I made was to use tuples rather than named { r, g, b } just for ease of typing the colors, I probably also want to add more diversity to how u can define colors by adding things like "from_hsl" "from_hsv" and if we are feeling extra fancy we can add these new cool-kid formats like "from_hwb" "from_lab" and "from_lch"

### CODE REVIEW

File: anathema-backend/src/tui/style.rs [hunk 1]
```diff
@@ -84,11 +89,19 @@ impl CellAttributes for Style {
...
-            Some(Color::Rgb { r, g, b }) => Some(Hex::from((r, g, b))),
+            Some(Color::Rgb(r, g, b)) => Some(Hex::from((r, g, b))),
             None | Some(_) => None,
         }
     }
 
+    fn get_color(&self, key: &str) -> Option<Color> {
+        match key {
+            "foreground" => self.fg,
+            "background" => self.fg,
```
Comment (FishingHacks) [inline anathema-backend/src/tui/style.rs:100]: i believe that `background => self.fg` is incorrect

```suggestion
            "background" => self.bg,
```

File: anathema-state/src/colors.rs [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+    LightMagenta,
+    LightCyan,
+    White,
+    /// 24bit color, expressed as rgb, following the spec
+    ///
+    /// See [24bit colors](https://en.wikipedia.org/wiki/ANSI_escape_code#24-bit)
+    Rgb(u8, u8, u8),
+    /// 8bit color.
+    ///
+    /// See [256 colorts](https://en.wikipedia.org/wiki/ANSI_escape_code#8-bit)
```
Comment (FishingHacks) [inline anathema-state/src/colors.rs:38]:
```suggestion
    /// See [256 colors](https://en.wikipedia.org/wiki/ANSI_escape_code#8-bit)
```

File: anathema-state/src/colors.rs [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+    fn from((r, g, b): (u8, u8, u8)) -> Self {
+        Self::Rgb(r, g, b)
+    }
+}
+
+impl TryFrom<&str> for Color {
+    type Error = ();
+
+    fn try_from(value: &str) -> Result<Self, Self::Error> {
+        Hex::try_from(value).map(Into::into)
```
Comment (FishingHacks) [inline anathema-state/src/colors.rs:64]: Shouldn't this use the FromStr trait as defined below?

File: anathema-widgets/src/widget/attributes.rs [hunk 4]
```diff
@@ -200,6 +200,15 @@ impl CellAttributes for Attributes<'_> {
         }
     }
 
+    fn get_color(&self, key: &str) -> Option<anathema_state::Color> {
+        let value = self.get_val(key)?;
+        let value = value.load_common_val()?;
+        match value.to_common()? {
+            CommonVal::Color(color) => Some(color),
```
Comment (FishingHacks) [inline anathema-widgets/src/widget/attributes.rs:207]:
```suggestion
            CommonVal::Color(color) => Some(color),
            CommonVal::Str(str) => Color::from_str(str).ok(),
            CommonVal::Int(ansi_value @ 0..=u8::MAX as i64) => Some(Color::AnsiVal(ansi_value as u8)),
```

The string case is handled by the TUI. I would suggest moving it down here with the code above. This way, lines 143-152 of `anathema-backend/src/tui/style.rs` can be simplified.

**Review by FishingHacks — CHANGES_REQUESTED** [review]

**Review by togglebyte — APPROVED** [review]

**Pull request merged**